src/clj/harja/kyselyt/kanavat/liikennetapahtumat.clj

```diff
@@ -51,49 +50,167 @@
+                            (when (and alku loppu)
+                              {::lt/aika (op/between alku loppu)})
+                            (when kohde-id
+                              {::lt/kohde-id kohde-id})
+                            (when toimenpide
+                              {::lt/sulku-toimenpide toimenpide})
+
+                            (op/and
+                              {::m/poistettu? false
+                               ::lt/urakka-id urakka-id
+                               ::lt/sopimus-id sopimus-id
+                               ::lt/kohde {::m/poistettu? false}}
+                              (when (or suunta aluslaji)
+                                {::lt/alukset (op/and
+                                                (when suunta
+                                                  {::lt-alus/suunta suunta})
+                                                (when aluslaji
+                                                  {::lt-alus/laji aluslaji}))})))))
+      (partial kanavat-q/hae-kohteiden-urakkatiedot db)
+      urakka-id)))
+
+(defn- hae-seuraava-kohde [db tiedot suunta]
+  nil)
+
+(defn- hae-kohteen-edellinen-tapahtuma* [tulokset]
+  (first
+    (sort-by ::lt/aika pvm/jalkeen?
+             tulokset)))
+
+(defn- hae-kohteen-edellinen-tapahtuma [db tapahtuma]
+  (let [urakka-id (::lt/urakka-id tapahtuma)
+        sopimus-id (::lt/sopimus-id tapahtuma)
+        kohde-id (::lt/kohde-id tapahtuma)]
+    (assert (and urakka-id sopimus-id kohde-id))
+    (hae-kohteen-edellinen-tapahtuma*
+      (specql/fetch
+        db
+        ::lt/liikennetapahtuma
+        (set/union
+          lt/perustiedot)
+        {::lt/kohde-id kohde-id
+         ::lt/urakka-id urakka-id
+         ::lt/sopimus-id sopimus-id}))))
+
+;; TODO Toteuta :)
+(defn- hae-kuittaamattomat-alukset [db kohde suunta]
+  nil)
+
+(defn hae-edelliset-tapahtumat [db tiedot]
+  ;; TODO kun kohteille tulee järjestys, niin täällä otetaan sekin huomioon
+  (let [ylos (hae-kuittaamattomat-alukset db tiedot :ylos)
+        alas (hae-kuittaamattomat-alukset db tiedot :alas)
+        kohde (hae-kohteen-edellinen-tapahtuma db tiedot)]
+    {:ylos ylos
+     :alas alas
+     :kohde kohde}))
+
+(defn- alus-kuuluu-tapahtumaan? [db alus tapahtuma]
+  (some?
+    (first
+      (specql/fetch db
+                    ::lt-alus/liikennetapahtuman-alus
+                    #{::lt-alus/id}
+                    {::lt-alus/liikennetapahtuma-id (::lt/id tapahtuma)
+                     ::lt-alus/id (::lt-alus/id alus)}))))
+
+(defn vaadi-alus-kuuluu-tapahtumaan! [db alus tapahtuma]
+  (assert (alus-kuuluu-tapahtumaan? db alus tapahtuma) "Alus ei kuulu tapahtumaan!"))
+
+(defn tallenna-alus-tapahtumaan! [db user alus tapahtuma]
+  (let [olemassa? (id-olemassa? (::lt-alus/id alus))
+        alus (assoc alus ::lt-alus/liikennetapahtuma-id (::lt/id tapahtuma))]
+    (if olemassa?
+      (do
+        (vaadi-alus-kuuluu-tapahtumaan! db alus tapahtuma)
```

Comment:
Tämä jäi vähän hämäräksi itselle. Eli siis kun päivitetään aluksen tietoja liikennetapahtumaan vaaditaan, että se liittyy jo entuudestaan siihen. Onko tilannetta, milloin näin ei olisi? Eli siis koitetaan päivittää tietoja alukselle tapahtumaan, johon se ei kuulu?

Reply:
Jos joku käsin kutsuu tuota palvelua, niin voi antaa alukselle id:ksi jotain häröä.